### PR TITLE
File-based discovery plugin

### LINKED ISSUE #20323: File based Unicast Host Provider
The unicast host list determines the list of port and ips an Elasticsearch node ping when it starts up and tries to find the cluster. The same list is used upon master lost and election start.

It has been long requested to be able to update this list via an API. That presented a challenge since, unlike other setting changes, this information can not be easily serialized in the cluster state as it needs to be used when the node, for example, is restarted.

An alternative is for us to develop a plugin that, similarly to the current EC2 discovery plugin, monitors a separate (configurable) file and supplies the list of port/ips stored in it when needed (i.e., when pinging starts).

The format for the file should be simple and contain one hostname/ip per line (in the same format the current yml file supports).

### PR DESCRIPTION
**Pull request by abeyad**
This PR introduces a new plugin for file-based unicast hosts
discovery. This allows specifying the unicast hosts participating
in discovery through a `discovery-file/unicast_hosts.txt` file located in the
config directory. The plugin will use the hosts specified in this
file as the set of hosts to ping during discovery.

The format of the `unicast_hosts.txt` file is to have one host/port
entry per line. The hosts file is read and parsed every time
discovery makes ping requests, thus a new version of the file that
is published to the config directory will automatically be picked
up.

Closes #20323

### CODE REVIEW

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Override
+    public List<DiscoveryNode> buildDynamicNodes() {
+        List<String> hostsList;
+        try {
+            hostsList = Files.readAllLines(unicastHostsFilePath);
+        } catch (FileNotFoundException | NoSuchFileException e) {
+            logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to find unicast hosts file [{}]",
+                                                                        unicastHostsFilePath), e);
+            return Collections.emptyList();
```
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: I don't know, do we have to have this leniency?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        } catch (IOException e) {
+            logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Error reading unicast hosts file [{}]",
+                                                                        unicastHostsFilePath), e);
+            return Collections.emptyList();
```
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: I really don't think we need this leniency, I'd like to understand why we're introducing it. I think we should just blow up the pings.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        final int hostsSize = hostsList.size();
+        for (int i = 0; i < hostsSize; i++) {
+            final String host = hostsList.get(i);
+            TransportAddress[] addresses;
+            try {
+                addresses = transportService.addressesFromString(host, 1);
+            } catch (Exception e) {
+                logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to parse transport address from [{}]",
+                                                                            host), e);
+                return Collections.emptyList();
```
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Same concern regarding the leniency.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            }
+            if (addresses.length != 1) {
+                logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to parse transport address from [{}]",
+                                                                            host));
+                return Collections.emptyList();
```
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Same concern regarding the leniency.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+            }
+            discoNodes.add(new DiscoveryNode("#zen_unicast_host_" + i + "#", addresses[0], Version.CURRENT.minimumCompatibilityVersion()));
+        }
+
+        logger.debug((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Using dynamic discovery nodes {}", discoNodes));
```
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: This can just be a plain old logging statement `logger.debug("[discovery-file] using dynamic discovery nodes {}", discoNodes);`.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    private final Settings settings;
+
+    public FileBasedDiscoveryPlugin(Settings settings) {
+        this.settings = settings;
+        logger.trace("starting file-based discovery plugin...");
+    }
+
+    public void onModule(DiscoveryModule discoveryModule) {
+        logger.debug("register file-based unicast hosts provider");
```
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java]: Maybe `trace`? And `register` -> `registering`?

File: plugins/discovery-file/src/test/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProviderTests.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    private MockTransportService transportService;
+
+    @BeforeClass
+    public static void createThreadPool() {
+        threadPool = new TestThreadPool(FileBasedUnicastHostsProviderTests.class.getName());
+    }
+
+    @AfterClass
+    public static void stopThreadPool() {
+        if (threadPool != null) {
```
Comment (jasontedor) [inline plugins/discovery-file/src/test/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProviderTests.java]: Maybe just `ESTestCase#terminate(threadPool);`

**Comment by dadoonet** [comment]
Just wondering why we are using a `txt` file here. I guess it makes it easier to be rendered by any 3rd party tool.
But can we imagine that we want to be able to add other settings in this file? So would it be better to generate a YAML format like:

```yaml
hosts:
  - host1:9300
  - host2
anyothersetting: foo
```

(Not an issue to me, just thinking loud)

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to parse transport address from [{}]",
+                                                                            host), e);
+                return Collections.emptyList();
+            }
+            if (addresses.length != 1) {
+                logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to parse transport address from [{}]",
+                                                                            host));
+                return Collections.emptyList();
+            }
+            discoNodes.add(new DiscoveryNode("#zen_unicast_host_" + i + "#", addresses[0], Version.CURRENT.minimumCompatibilityVersion()));
```
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Maybe it should have file in the name? And maybe this shouldn't be simply incrementing on each round because it means from one round to the next we're reusing names and maybe they should be unique?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        super(settings);
+        this.transportService = transportService;
+        this.unicastHostsFilePath = new Environment(settings).configFile().resolve(UNICAST_HOSTS_FILE);
+    }
+
+    @Override
+    public List<DiscoveryNode> buildDynamicNodes() {
+        List<String> hostsList;
+        try {
+            hostsList = Files.readAllLines(unicastHostsFilePath);
```
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Doesn't this mean we're going to be re-reading and parsing the file on every round of pinging, even if nothing has changed?

File: plugins/discovery-file/build.gradle [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ * under the License.
+ */
+
+esplugin {
+    description 'Discovery file plugin enables unicast discovery from hosts stored in a file.'
+    classname 'org.elasticsearch.discovery.file.FileBasedDiscoveryPlugin'
+}
+
+// no integration tests for this plugin
+integTest.enabled = false
```
Comment (dadoonet) [inline plugins/discovery-file/build.gradle]: Can we add a simple REST test like this one? https://github.com/elastic/elasticsearch/blob/master/plugins/discovery-ec2/src/test/resources/rest-api-spec/test/discovery_ec2/10_basic.yaml

File: plugins/discovery-file/src/test/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProviderTests.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.hamcrest.Matchers.containsString;
+
+/**
+ * Tests for {@link FileBasedUnicastHostsProvider}.
+ */
+public class FileBasedUnicastHostsProviderTests extends ESTestCase {
```
Comment (jasontedor) [inline plugins/discovery-file/src/test/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProviderTests.java]: However we decide to handle the error cases (with the existing leniency, or throwing an exception), shouldn't there be a test for those code paths?

**Comment by dadoonet** [comment]
I think we need to add it also here: 

* https://github.com/elastic/elasticsearch/blob/master/qa/vagrant/src/test/resources/packaging/scripts/50_modules_and_plugins.bats

* https://github.com/elastic/elasticsearch/blob/master/qa/vagrant/src/test/resources/packaging/scripts/module_and_plugin_test_cases.bash
* https://github.com/elastic/elasticsearch/blob/master/dev-tools/smoke_test_rc.py

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 4 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: I'm also not terribly comfortable with the leniency, but it was done this way for two reasons:
 1. Its more consistent with how the other unicast host providers deal with exceptions (e.g. https://github.com/elastic/elasticsearch/blob/master/plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2UnicastHostsProvider.java#L120 and https://github.com/elastic/elasticsearch/blob/master/plugins/discovery-gce/src/main/java/org/elasticsearch/discovery/gce/GceUnicastHostsProvider.java#L251)
 2. Second, its possible that the hosts file will publish at the same time as we are reading it, possibly causing IO exceptions, so the idea was to just return an empty list and we pick up the file in the next round.

That said, I'd be in favor of throwing exceptions here and in the other places instead.

File: plugins/discovery-file/src/test/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProviderTests.java (hunk 7 above)
Comment (abeyad) [inline plugins/discovery-file/src/test/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProviderTests.java]: done

File: plugins/discovery-file/src/test/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProviderTests.java (hunk 11 above)
Comment (abeyad) [inline plugins/discovery-file/src/test/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProviderTests.java]: That was kind of the idea with `testUnicastHostsDoesNotExist`.  The problem with testing the validity of host entries is that I was using a local mock transport service, so it uses `LocalTransport` instead of `TcpTransport`.  `LocalTransport` does not try to parse the host entries but rather just uses whatever the string is as the transport address id. I've switched to use a mock transport service with `MockTcpTransport`, so now I can test the scenario when given a bad host entry.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 5 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Done

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 9 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Yes, we are re-reading the file on each round.  @bleskes preferred it this way for now for simplicity.  We can change it to use a `WatchService` on the file if performance is a concern.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java (hunk 6 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java]: done

File: plugins/discovery-file/build.gradle (hunk 10 above)
Comment (abeyad) [inline plugins/discovery-file/build.gradle]: good point, added

**Comment by abeyad** [comment]
> I think we need to add it also here:

@dadoonet thanks for alerting me to these!

**Comment by abeyad** [comment]
> Just wondering why we are using a txt file here.

@dadoonet the idea was to keep it really simple here.  Since this meant to be for machine generated files and not user facing (otherwise, they would've entered the hosts in `elasticsearch.yaml`), the yaml format doesn't really add much value and a simple text file of one entry per line is easy to output and easy to read in (and also easy for anyone to read who wants to).

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 8 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Doen

**Comment by abeyad** [comment]
I pushed a few commits to address code review comments, except for the leniency issue and the file re-reading issue pending feedback and a decision on that.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 9 above)
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: since we ping every 3 second at most and only upon master loss/start up - I don't see why we need the added complexity of a watch? I think we can do the simplest thing possible here, like what Ali did?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 1 above)
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: We expect people to concurrently update this file, so we have to deal with concurrent modifications or that we won't be able to access the file, find it half written, deleted (to be re-written) etc. On start up we can maybe fail but if it happens just before master loss, it's harder.  I think the simplest thing here is to WARN log and return an empty list? this would mean we ping all existing nodes in the last known cluster state. If that's enough for the election, fine. If not, we'll go re-pinging and try to read the file again. 

Any other suggestions?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 4 above)
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: > That said, I'd be in favor of throwing exceptions here and in the other places instead

See my comment above. We can throw exceptions here, but we'll just catch it higher up (and log it). We also ping based on the cluster state, so we can let the pinging continue and see how it goes.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 9 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: To recap a discussion between @bleskes, @jasontedor, and myself, the main concern here was potentially needing to resolve DNS on each round if host names are used. The conclusion is that we are as yet uncertain about how to deal with the case of re-resolving DNS and how to deal with slow DNS servers.  We have the same issue now with the EC2 discovery plugin, for example, if EC2 is non-responsive for whatever reason.  There are many factors to consider here, so in light of that, we will stick with the simplest solution for now and just re-read and parse the file on each ping round.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 1 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Upon discussion between @bleskes, @jasontedor, and myself, we will stick with the current implementation of returning an empty list with a `WARN` log message.

**Comment by abeyad** [comment]
retest this please

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+    public FileBasedDiscoveryPlugin(Settings settings) {
+        this.settings = settings;
+        logger.trace("starting file-based discovery plugin...");
+    }
+
+    public void onModule(DiscoveryModule discoveryModule) {
+        logger.trace("registering file-based unicast hosts provider");
+        discoveryModule.addDiscoveryType(DISCOVERY_FILE, ZenDiscovery.class);
+        discoveryModule.addUnicastHostProvider(DISCOVERY_FILE, FileBasedUnicastHostsProvider.class);
```
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java]: I don't think we need a dedicate discovery type? I think we can just use `zen`?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 1 above)
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: can we not short cut return, but rather set the hostList to empty and continue the same execution path (with the same logging)

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+            TransportAddress[] addresses;
+            try {
+                addresses = transportService.addressesFromString(host, 1);
+            } catch (Exception e) {
+                logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to parse transport address from [{}]",
+                                                                            host), e);
+                return Collections.emptyList();
+            }
+            if (addresses.length != 1) {
+                logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Failed to parse transport address from [{}]",
```
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: in what cases do we end up here, without an exception in the earlier `transportService.addressesFromString(host, 1);` clause?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 3 above)
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: if we hit an exception here, why do we give up on the rest? I think we should use what we can?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+            return Collections.emptyList();
+        } catch (IOException e) {
+            logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Error reading unicast hosts file [{}]",
+                                                                        unicastHostsFilePath), e);
+            return Collections.emptyList();
+        }
+
+        final List<DiscoveryNode> discoNodes = new ArrayList<>();
+        final int hostsSize = hostsList.size();
+        for (int i = 0; i < hostsSize; i++) {
```
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: can't we just use a simple foreach?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 1 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Done

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 14 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Done

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 3 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: We can use what we can, but I thought returning an empty list would be a quicker indication that there was an entry in the file that we could not get an address from (either not parsable, or DNS down, etc). If we return all but one entry, I thought it may be harder to decipher that faulty entry in the unicast hosts file.  If we keep returning an empty list and we aren't able to converge on an election as a result, it would alert the user to the faulty entries in the file much quicker.  It would seem to me the likely scenario of having one faulty entry and the rest sound entries is that the file was written incorrectly and the user of the plugin would need to change something on their end to fix it.

I could go either way on this, I just wanted to explain my reasoning for why I chose to return an empty list if we encounter one faulty entry.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 13 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: It should be impossible, I will remove this redundant check

**Comment by bleskes** [comment]
I went through this and left some minor comments.

I think it will be great if:
1) Add a default config which is installed with the plugin
2) Support comments i.e., remove lines that start with `#`.
3) Populate that file with a header that explains what it's doing there.

As follow ups, I would love to see:
1) Documentation
2) A proper integration test where the plugin is formed based on information in the config file. This way we really know it works.

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 3 above)
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: I see what you are saying I would agree with you if we fail to start because of this. However, when it comes to a live and running node I think it will be surprising (to me) if we would use what we can due to single entry. It's kind of a mixture of being tolerant and strict at the same time.

Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Sounds good, i'll make the change

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java (hunk 12 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedDiscoveryPlugin.java]: I changed it to use `zen` and just add a unicast host provider to `zen`

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 3 above)
Comment (jasontedor) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: I agree with @abeyad's reasoning, this sounds even worse than just returning an empty list. 😦

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+        } catch (IOException e) {
+            logger.warn((Supplier<?>) () -> new ParameterizedMessage("[discovery-file] Error reading unicast hosts file [{}]",
+                                                                        unicastHostsFilePath), e);
+            hostsList = Collections.emptyList();
+        }
+
+        final List<DiscoveryNode> discoNodes = new ArrayList<>();
+        for (final String host : hostsList) {
+            TransportAddress[] addresses;
+            try {
```
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: can we do what UnicastZenPing does? Note how it iterates the addresses

```
                TransportAddress[] addresses = transportService.addressesFromString(host, limitPortCounts);
                for (TransportAddress address : addresses) {
                    configuredTargetNodes.add(new DiscoveryNode(UNICAST_NODE_PREFIX + unicastNodeIdGenerator.incrementAndGet() + "#",
                            address, emptyMap(), emptySet(), getVersion().minimumCompatibilityVersion()));
                }

```

better yet - can we make this a static method in UnicastZenPing and use it here?

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+    private final Path unicastHostsFilePath;
+
+    private final AtomicLong nodeIdGenerator = new AtomicLong(); // generates unique ids for the node
+
+    @Inject
+    public FileBasedUnicastHostsProvider(Settings settings, TransportService transportService) {
+        super(settings);
+        this.transportService = transportService;
+        this.unicastHostsFilePath = new Environment(settings).configFile().resolve("discovery-file").resolve(UNICAST_HOSTS_FILE);
```
Comment (bleskes) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: why did you add this `discovery-file` directory?

Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: This is the convention for plugins to have their own directories for their own config files within `config`

File: plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java (hunk 15 above)
Comment (abeyad) [inline plugins/discovery-file/src/main/java/org/elasticsearch/discovery/file/FileBasedUnicastHostsProvider.java]: Sure I'll do that.  I didn't do the iteration since there should only be 1 returned address (limitPortcounts = 1)

**Comment by bleskes** [comment]
LGTM. Thx @abeyad . Please test manually (to cover for the lack of a full blown integration test) and open issues for the follow ups we discussed.

**Comment by abeyad** [comment]
Thanks for the review @bleskes !

**Comment by abeyad** [comment]
I added https://github.com/elastic/elasticsearch/pull/20394/commits/171309558c5dd15b7454aae117cba5e8b7690b88 to support (very basic) having comments in the unicast_hosts.txt file

**Comment by abeyad** [comment]
This final commit on adding support for comments in `unicast_hosts.txt` looks good to @dakrone, so I'm merging.

**Comment by abeyad** [comment]
5.0 commit: ff9058018da1e0f2f0bd8a765b5446499e3811ef
5.x commit: f66b59af51d103f90428c7c023ea63294041aee9

**Comment by abeyad** [comment]
5.0 commit: ff9058018da1e0f2f0bd8a765b5446499e3811ef
5.x commit: f66b59af51d103f90428c7c023ea63294041aee9
